R/build_xgboost.R

```diff
@@ -629,3 +745,627 @@ glance.xgb.Booster <- function(x, pretty.name = FALSE, ...) {
   }
   ret
 }
+
+# XGBoost prediction function that takes data frame rather than matrix.
+predict_xgboost <- function(model, df) {
+  y_name <- all.vars(model$terms)[[1]]
+  if(is.null(df[[y_name]])){
+    # if there is no column in the formula (even if it's response variable),
+    # model.matrix function causes an error
+    # so create the column with 0
+    df[[y_name]] <- rep(0, nrow(df))
+  }
+
+  mat_data <- if(!is.null(model$is_sparse) && model$is_sparse){
+    Matrix::sparse.model.matrix(model$terms, data = model.frame(df, na.action = na.pass, xlev = model$xlevels))
+  } else {
+    model.matrix(model$terms, model.frame(df, na.action = na.pass, xlev = model$xlevels))
+  }
+
+  predicted <- stats::predict(model, mat_data)
+
+  obj <- model$params$objective
+
+  if (obj == "multi:softprob") {
+    # predicted is a vector containing probabilities for each class
+    probs <- matrix(predicted, nrow = length(model$y_levels)) %>% t()
+    # probs <- fill_mat_NA(as.numeric(rownames(mat_data)), probs, nrow(df)) # code from xgboost step.
+    probs <- fill_mat_NA(1:nrow(df), probs, nrow(df)) # Not sure if this is necessary. TODO: check.
+    colnames(probs) <- model$y_levels
+    predicted <- probs
+    # We return matrix in this case.
+    # Converting it to data.frame is the caller's responsibility.
+  }
+  else if (obj == "multi:softmax") {
+    predicted <- model$y_levels[predicted+1]
+    # fill rows with NA
+    predicted <- fill_vec_NA(as.integer(rownames(mat_data)), predicted, nrow(df))
+  }
+  else { # TODO: Here we assume all other cases returns vector prediction result.
+    # model.matrix removes rows with NA and stats::predict returns a matrix
+    # whose number of rows is the same with its size,
+    # so the result should be filled by NA
+    predicted <- fill_vec_NA(as.integer(rownames(mat_data)), predicted, nrow(df))
+  }
+  predicted
+}
+
+
+partial_dependence.xgboost <- function(fit, vars = colnames(data),
+  n = c(min(nrow(unique(data[, vars, drop = FALSE])), 25L), nrow(data)),
```

Comment:
How can I change 25? Probably we should have another argument for max number of categories?